libs/gltfio/include/gltfio/math.h

```diff
@@ -66,8 +67,17 @@ UTILS_PUBLIC inline void decomposeMatrix(const filament::math::mat4f& mat, filam
     float scaley = length(float3({d, e, f}));
     float scalez = length(float3({g, h, i}));
     float3 s = { scalex, scaley, scalez };
-    if (det < 0) {
-        s = -s;
+    if (preservedScale != float3{}) {
+        const float signx = (preservedScale.x < 0) ? -1 : 1;
```

Comment:
You are using an entire float3 to store 3 bits of information. All you need is whether each of x/y/z was negative or not.